### ISSUE TITLE
host_rule will always be reordered

### ISSUE THREAD

**Issue by merceskoba** [hashicorp/terraform-provider-google#7859]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version

Terraform v0.13.5
+ provider registry.terraform.io/hashicorp/google v3.48.0

### Affected Resource(s)

google_compute_url_map

### Terraform Configuration Files

```tf
resource "google_compute_url_map" "url_map_a" {
  project = "playground-public"
  name = "url-map-a"
  default_service = google_compute_backend_service.playground_a.id

  host_rule {
    hosts = ["sdwr.domain.tld"]
    path_matcher = "spaths"
  }

  host_rule {
    hosts = ["beta.domain.tld"]
    path_matcher = "bpaths"
  }

  path_matcher {
    name = "spaths"
    default_service = google_compute_backend_service.playground_sbs.id
  }

  path_matcher {
    name = "bpaths"
    default_service = google_compute_backend_service.playground_bbs.id
  }
```

TLDR;

1. `terraform apply`
```
- host_rule {
          - hosts        = [
              - "sdwr.domain.tld",
            ] -> null
          - path_matcher = "spaths" -> null
   }
+ host_rule {
          + hosts        = [
              + "beta.domain.tld",
            ]
          + path_matcher = "bpaths"
    }
 + host_rule {
          + hosts        = [
              + "sdwr.domain.tld",
            ]
          + path_matcher = "spaths"
    }
```

TL;DR
Every part of new host_rule i add, the terraform will re-sort the host_rule order. It will make the terraform looks like removing existing host_rule and then add back the existing host_rule and the new host_rule in the same time.
I am affraid to use this in production, because i have 39 host_rule.
I cannot imagine if i add the 40th host_rule and terraform shows it removes 39 host_rule and then add 39 host_rule back + 1 new host_rule...
Magicly, when i apply it, the existing hosts aren't removed. So, it's like the terraform removes them only in terraform side but not in google cloud side

**Comment by edwardmedia** [comment]
@ndmckinley shouldn't it be TypeList instead?

https://github.com/hashicorp/terraform-provider-google/blob/ee251c46904022e68f857408f2192295d3f18ecf/google/resource_compute_url_map.go#L687

**Comment by merceskoba** [comment]
Could we push the fix in next release, please ?

**Comment by ndmckinley** [comment]
Ah, that's a really unfortunate diff - that should be a lot more readable than it is.

Unfortunately, I believe this behavior is correct to the extent we can control it in the provider.  Because this is a set, the order is not preserved / not important, and you are correct - the host rules are _not_ removed at any step of the process.  The update you're proposing should be safe to apply.

If the main issue that you're having is that it is difficult to read diffs of sets of objects, that's definitely worth fixing, but unfortunately that's a terraform core issue and I believe you'll be better off filing it in github.com/hashicorp/terraform, since we cannot change anything about core terraform behavior.

If someone over there needs help debugging, feel free to point them to me, but I'll close this issue because I don't think it's something we can fix in our provider.

**Comment by merceskoba** [comment]
I got it, @ndmckinley 
I just check [named_port](https://github.com/hashicorp/terraform-provider-google/blob/master/google/resource_compute_instance_group_manager.go#L120)
It's a TypeSet because host_rule are push togather at the same time, like named_port do
Thank you for enlightening me

**Comment by antrusd** [comment]
@ndmckinley , what would be the problem if `host_rule` changed to TypeList rather than keep as TypeSet ?

CMIIW, if we use TypeList, we can move the updated `host_rule` to the bottom of List then updating or removing should not change the rest.

**Comment by ndmckinley** [comment]
It would show a diff if the API returned the list in a different order than the one you maintain.  Since the API doesn't guarantee any particular order, that would probably happen a lot, maybe every time.
